node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1415,34 +1412,29 @@ impl BlockSignaturesCollector {
     }
 }
 
-/// In the "sync to genesis" process, we need to consider finality signatures as valid when
-/// their total weight is at least `finality_threshold_fraction` of the total validator weights.
-///
-/// Please note that this is a relaxed requirement compared to the normal (non "sync to genesis")
-/// chain operation where we use the [quorum fraction](consensus::utils::quorum_fraction) calculated
-/// from the `finality_threshold_fraction` parameter.
-fn check_sufficient_finality_signatures_for_sync_to_genesis(
-    validator_weights: &BTreeMap<PublicKey, U512>,
-    finality_threshold_fraction: Ratio<u64>,
-    sigs: &BlockSignatures,
-) -> Result<(), FinalitySignatureError> {
-    consensus::check_sufficient_finality_signatures_with_quorum_formula(
-        validator_weights,
-        finality_threshold_fraction,
-        sigs,
-        std::convert::identity,
-    )
-}
-
-// Returns true if the output from consensus can be interpreted
-// as sufficient finality signatures for the sync to genesis process.
-fn are_signatures_sufficient_for_sync_to_genesis(
-    consensus_verdict: Result<(), FinalitySignatureError>,
-) -> bool {
-    match consensus_verdict {
-        Ok(_) | Err(FinalitySignatureError::TooManySignatures { .. }) => true,
-        Err(_) => false,
-    }
+/// Reads the validator weights that should be used to check the finality signatures for the given
+/// block.
+async fn era_validator_weights_for_block<'a>(
+    block_header: &BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<(EraId, BTreeMap<PublicKey, U512>), Error> {
+    let era_for_validators_retrieval = get_era_id_for_validators_retrieval(
+        &block_header.era_id(),
+        ctx.config.last_emergency_restart(),
+    );
+    let switch_block_of_previous_era = ctx
+        .effect_builder
+        .get_switch_block_header_at_era_id_from_storage(era_for_validators_retrieval)
+        .await
+        .ok_or(Error::NoSwitchBlockForEra {
+            era_id: era_for_validators_retrieval,
+        })?;
+    let validator_weights = switch_block_of_previous_era
+        .next_era_validator_weights()
+        .ok_or(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
```

Comment:
Is that really the right error variant? Does `get_switch_block_header_at_era_id_from_storage` ever return a non-switch block?

Reply:
Yes, I was thinking about it in the other PR: https://github.com/casper-network/casper-node/pull/3125#discussion_r905843435

Similarly to [this](https://github.com/casper-network/casper-node/pull/3142#discussion_r907325670), ideally `get_switch_block_header_at_era_id_from_storage()` should return new type `Option<SwitchBlockHeader>`, which should have the `next_era_validator_weights()` returning non-option type.

For now I think I'll follow the suggestion from @goral09 linked above.